util/stmtsummary/statement_summary.go

```diff
@@ -414,228 +479,311 @@ func newStmtSummaryByDigest(sei *StmtExecInfo, beginTime int64) *stmtSummaryByDi
 	tableNames := buffer.String()
 
 	ssbd := &stmtSummaryByDigest{
-		beginTime:     beginTime,
 		schemaName:    sei.SchemaName,
 		digest:        sei.Digest,
 		stmtType:      strings.ToLower(sei.StmtCtx.StmtType),
 		normalizedSQL: normalizedSQL,
 		tableNames:    tableNames,
-		indexNames:    strings.Join(sei.StmtCtx.IndexNames, ","),
-		minLatency:    sei.TotalLatency,
-		backoffTypes:  make(map[fmt.Stringer]int),
-		firstSeen:     sei.StartTime,
-		lastSeen:      sei.StartTime,
+		history:       list.New(),
 	}
-	ssbd.add(sei)
+	ssbd.add(sei, beginTime, historySize)
 	return ssbd
 }
 
-func (ssbd *stmtSummaryByDigest) add(sei *StmtExecInfo) {
+func (ssbd *stmtSummaryByDigest) add(sei *StmtExecInfo, beginTime int64, historySize int) {
+	// Enclose this block in a function to ensure the lock will always be released.
+	ssElement, ok := func() (*stmtSummaryByDigestElement, bool) {
+		ssbd.Lock()
+		defer ssbd.Unlock()
+
+		var ssElement *stmtSummaryByDigestElement
+		ok := false
```

Comment:
The logic is tricky here, please rename this variable and add some comments

Reply:
Addressed.